proposals/11991-tests-refactor/README.md

```diff
@@ -0,0 +1,207 @@
+
+#### Full Exploratory Tests
+Using an AI tool, we can compile a pipeline and validate it against the Pipeline Upload API (so basically, leverage Semi Exploratory tests but use an AI tool to generate a YAML):
+![SDK testing - Full Exploratory.png](SDK%20testing%20-%20Full%20Exploratory.png)
+
+**Pros**:
+* Fully self managed tests
+* Provides much broader coverage of edge/non edge cases that we may not have covered in targeted/semi targeted tests
+
+**Cons**
+* We may have to build a custom SDG - out of the box models may not work for us
+* Potentially duplicate coverage
+* Fail on edge cases that are of trivial priority
+
+### End to End Tests
+#### Critical Regression Testing
+Using a subset of the main list of pipeline files (already used for API tests and SDK tests), we can run full end to end tests that will confirm the integration of different components that make up KFP.
+![End2End-DuringPR.png](End2End-DuringPR.png)
+
+#### Full Regression Testing
+Develop a simple application to run pipeline files from a directory (local or remote), run them in parallel, and verify successful execution.
+![End2End-Full Regression.png](End2End-Full%20Regression.png)
+
+**Pros**:
+An ad hoc way for us to load the system
+* Full end to end testing
+* Lot of indirect testing of components without really worrying about the internals
+* Easy to expand - all it will take is to generate a new pipeline file
+* Flexible testing app for other products that integrates with KFP
+* Easy integration with AI tools to perform some exploratory testing
+
+**Cons**:
+* Failed pipelines may get hard to debug
+* Resource utilization will be high
+* Some pipelines may not be possible due to resource constraints
+* Long running tests (even after parallelization)
+* Parallel tests can cause storage to fill up on the GitHub workers, since we clean up only after the pipeline run finishes. So we may need to explore running tests in parallel on separate GitHub workers/workflows.
+
+## Test Architecture Changes
+* Implement all tests in Go using Ginkgo + Gomega, providing BDD-style, readable, and organized test scripts with multi-format output and parallel execution.
+* Use Pytest for SDK tests
+* Sample tests will be renamed to End to End Tests and labeled with “Critical” label. These tests will be a subset of full regression tests covering critical use cases, and will run with every PR to confirm regression in integration. These tests will also be converted to Ginkgo tests.
+* Full Regression (Lightweight End-to-end) Tests will run on a schedule May be start with every merge to master and then change frequency as we gain confidence and as the list of tests grow, to daily, weekly or per release. Go.mod file will be added to these tests to allow reuse of this code at other places.
```

Comment:
I think we should stick to the same tests running on PR CI and merges. If it starts to take too long for PR CI, we can reevaluate later.